### DISCUSSION TITLE
Camera support (General)

### DISCUSSION THREAD

**Discussion by StollD** [linux-surface/linux-surface#1355] [Cameras]
This is a general discussion space for camera support on surface devices. You can use this if you want to ask broad, general questions that don't warrant their own ticket, or you want to share news / your own work (i.e. scripts).

**libcamera is as low-level as it should get here! For hardware specific discussion, use the following topics:**
* https://github.com/linux-surface/linux-surface/discussions/1352
* https://github.com/linux-surface/linux-surface/discussions/1353
* https://github.com/linux-surface/linux-surface/discussions/1354

**If you have a specific issue, please DO NOT post here. Open an issue on the issue tracker instead.**

**Comment by ikshovon** [comment]


**Reply by StollD** [reply]
From the OP:

> **If you have a specific issue, please DO NOT post here. Open an issue on the issue tracker instead.**

Please open a new issue with how you are trying to use the cameras, as well as the output of the `dmesg` command.

**Comment by g3blv** [comment]
How can I find out what camera sensor the Surface Go 4 has? I haven't been able to find any information about it.

**Comment by LMerbecks** [comment]
# Support Confusion

I am confused about the [Kernel support](https://github.com/linux-surface/linux-surface/wiki/Camera-Support#kernel-support) and [libcamera support](https://github.com/linux-surface/linux-surface/wiki/Camera-Support#libcamera-support) sections in the Camera Support wiki page.

- Do the kernel binaries contain the camera patches already?
- Is libcamera needed for anything else other than testing the Surface cameras?

For me on a Surface Pro 6 running `6.17.1-surface-2` with ubuntu `24.04` the cameras do not work when I try to use them from the Camera Ubuntu app. However, I have libcamera installed and can access the video feed from `qcam` even though I get this error message:

``` 
QSocketNotifier: Can only be used with threads started with QThread
[1:12:55.840279955] [21281]  INFO Camera camera_manager.cpp:284 libcamera v0.2.0
[1:12:55.990222365] [21293] ERROR IPAProxy ipa_proxy.cpp:149 Configuration file 'ov8865.yaml' not found for IPA module 'ipu3'
[1:12:55.990844190] [21293]  INFO IPU3 ipu3.cpp:1141 Registered Camera[0] "\_SB_.PCI0.I2C3.CAMR" connected to CSI-2 receiver 0
[1:12:55.994525857] [21293] ERROR IPAProxy ipa_proxy.cpp:149 Configuration file 'ov5693.yaml' not found for IPA module 'ipu3'
[1:12:55.995104792] [21293]  INFO IPU3 ipu3.cpp:1141 Registered Camera[1] "\_SB_.PCI0.I2C2.CAMF" connected to CSI-2 receiver 1
```

Because of the [Support table](https://github.com/linux-surface/linux-surface/wiki/Camera-Support#libcamera-support) containing check marks for the SP6 I am expecting the cameras to work. The SP6 row was, however, already checked in the first revision of the Camera Support page committed by @xz-dev. Did cameras work at that time already and what were the steps to get there?

The [Surface Pro 6](https://github.com/linux-surface/linux-surface/wiki/Surface-Pro-6) wiki page is also not helpful as it only covers thermal problems while running on CachyOS.

I am ready to get RTFM'ed and have no idea if this issue or [this](https://github.com/linux-surface/linux-surface/discussions/1352) is the correct one to post this but at the same time also want to contribute by documenting the setup more clearly.

**Reply by kbingham** [reply]
Yes, as SP6 is an IPU3 I expect it to be possible to use in modern desktop set ups with Pipewire and up to date Firefox and Chrome browsers.

libcamera *is* required as it is responsible for performing the ISP configuration and 3a. It should all work as you've seen with Qcam, but unfortunately is likely a bit green - as  these cameras have never been calibrated or tuned. (Hence the error message stating no configuration file is available) - but even without those it should give you a visible picture.

**Reply by LMerbecks** [reply]
Alright, I remembered just now, that the [feature matrix](https://github.com/linux-surface/linux-surface/wiki/Supported-Devices-and-Features) also states in the footnote of the SP6 that libcamera is required.

Should I, however, expect the cameras to be working in e.g. the camera app? Because they do not right now and I am stuck on how to proceed. 

As for the qcam images: I would say mine are more into the red part for the front camera and maybe a bit green-ish for the back camera. The quality is decent. However, the back camera is flipped vertically. 

Also, I saw [here](https://github.com/linux-surface/linux-surface/discussions/1352#discussioncomment-13923660) the camera's ISP might need tuning. Sounds like a fun project tbh and there is some very crude [documentation](https://docs.kernel.org/admin-guide/media/ipu3.html) on the IPU3 ISP parameters, but I would not even know where to start testing/implementing changes to the ISP parameters.

**Reply by kbingham** [reply]
Tuning is done at the libcamera level - rather than the ISP level - so you don't have to know the internal details of the ISP parameters.

Raspberry Pi have a great write up on how it's done : https://datasheets.raspberrypi.com/camera/raspberry-pi-camera-guide.pdf

That's specific to Raspberry Pi - but the other platforms either do or will follow very closely. There are still a few more pieces waiting to get merged to be able to easily tune the IPU6 cameras through the SoftISP though - the RAW stream support I hope will be merged quite soon which will unblock a lot of this, but there's already things that you could get into if you have time and motivation to explore the topic.

**Reply by LMerbecks** [reply]
I'm looking into it.

**Comment by kbingham** [comment]
> Should I, however, expect the cameras to be working in e.g. the camera app? Because they do not right now and I am stuck on how to proceed.

To answer that can you be more specific about what the "camera app" is ? QCam is "a camera app" and you've already confirmed the camera works there!

**Reply by LMerbecks** [reply]
Sure thing, I will remember that for the future :)
The camera app in question is gnome camera (`snapshot` package). But I would also expect programs like zoom (roast me) to be able to access the camera natively with the surface kernel. 

If I launch `snapshot` I get the no camera found icon:
<img width="1844" height="1524" alt="Camera not found image" src="https://github.com/user-attachments/assets/ec066a1d-9c3d-481b-9252-65e01b5232e6" />

Triggered by your follow up question, I managed to get the camera working in snapshot briefly by following [this answer](https://askubuntu.com/a/1515774). TL;DR: 
```console
~ sudo apt update
~ systemctl --user restart pipewire
~ snapshot
```

After this snapshot works. Zoom for example, however, still does not find the cameras. Moreover, after a reboot the cameras are not found again. 

Note I am oblivious to the commands above beyond the fact that
1. we update all packages
2. we restart the pipewire service (which you already mentioned and I get is a service that handles video and audio processing)

**Reply by derkrasseleo** [reply]
can't confirm with SP9, though that has ipu6

**Reply by LMerbecks** [reply]
> can't confirm with SP9, though that has ipu6

@derkrasseleo can you be more specific what you mean by that? The `pipewire` restart did not fix anything or what?

**Reply by derkrasseleo** [reply]
> > can't confirm with SP9, though that has ipu6
> 
> @derkrasseleo can you be more specific what you mean by that? The `pipewire` restart did not fix anything or what?

Exactly, I'm also very confused on how to debug the surface camera issue, could you maybe tell me which versions of pipewire, libcamera etc. you are running?

**Reply by LMerbecks** [reply]
Sure. I'm running
- pipewire 1.0.5
- libcamera  0.2.0-3fakesync1build6
- surface kernel 6.17.1-surface-2
on ubuntu 24.04.3 LTS. 

Unfortunately, I am kind of in the same position as you are, as I have hardly any clue on how to debug the camera properly. Though if you want to get into the low level stuff I suggest you read (at least the last comment) on #1354. There is some activity there.
